vertical-pod-autoscaler/pkg/utils/status/status_object.go

```diff
@@ -0,0 +1,165 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"net"
+	"time"
+
+	apicoordinationv1 "k8s.io/api/coordination/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	typedcoordinationv1 "k8s.io/client-go/kubernetes/typed/coordination/v1"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	// Parameters for retrying with exponential backoff.
+	retryBackoffInitialDuration = 100 * time.Millisecond
+	retryBackoffFactor          = 3
+	retryBackoffJitter          = 0
+	retryBackoffSteps           = 6
+)
+
+// Client for the status object.
+type Client struct {
+	client               typedcoordinationv1.LeaseInterface
+	name                 string
+	namespace            string
+	leaseDurationSeconds int32
+	holderIdentity       string
+}
+
+// NewClient returns a client for the status object.
+func NewClient(c clientset.Interface, name, namespace string, leaseDuration time.Duration, holderIdentity string) *Client {
+	return &Client{
+		client:               c.CoordinationV1().Leases(namespace),
+		name:                 name,
+		namespace:            namespace,
+		leaseDurationSeconds: int32(leaseDuration.Seconds()),
+		holderIdentity:       holderIdentity,
+	}
+}
+
+// UpdateStatus renews status object lease.
+// Status object will be created if it doesn't exist.
+func (c *Client) UpdateStatus() error {
+	updateFn := func() error {
+		lease, err := c.client.Get(c.name, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			// Create lease if it doesn't exist.
+			return c.create()
+		} else if err != nil {
+			return err
+		}
+		lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+		_, err = c.client.Update(lease)
+		if apierrors.IsConflict(err) {
+			// Lease was updated by an another replica of the component.
+			// No error should be returned.
+			return nil
+		}
+		return err
+	}
+	return retryWithExponentialBackOff(updateFn)
+}
+
+func (c *Client) create() error {
+	_, err := c.client.Create(c.newLease())
+	if apierrors.IsAlreadyExists(err) {
+		// Lease was created by an another replica of the component.
+		// No error should be returned.
+		return nil
+	}
+	return err
+}
+
+func (c *Client) newLease() *apicoordinationv1.Lease {
+	return &apicoordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.name,
+			Namespace: c.namespace,
+		},
+		Spec: apicoordinationv1.LeaseSpec{
+			HolderIdentity:       pointer.StringPtr(c.holderIdentity),
+			LeaseDurationSeconds: pointer.Int32Ptr(c.leaseDurationSeconds),
+		},
+	}
+}
+
+// GetStatus returns status object.
+func (c *Client) GetStatus() (*apicoordinationv1.Lease, error) {
```

Comment:
An alternative to retrying getting the lease object 6 times in GetStatus() is simply skipping actuation in updater if GetStatus fails. We will retry in 1 minute anyway.

Reply:
I agree, but I wanted to have generic module for status. Plus, retries are always a good idea.
Maybe 6 time is overkill, I will change it to 3...

Reply:
Can't agree retries are always a good idea. The retryWithBackoff function is used here for GetStatus and for UpdateStatus, which will be called every 10 seconds, right? Removing retrywithbackoff simplifies things a lot here. 
Might be worth considering. I won't be very stuborn if you want to keep it though.

Reply:
> Removing retrywithbackoff simplifies things a lot here.

I agree.

> GetStatus and for UpdateStatus, which will be called every 10 seconds, right?

UpdateStatus will be called every 10 seconds. if no update was made for 1m we assume that the component is not available. Therefore we have ~6 six attempts to update the status. Retrying every single update doesn't seem to be necessary.
The question is what approach we have for GetStatus?... If GetStatus fails, does it means that component is not available? It needs to be retried somehow. There are three solutions that I see:
-  GetStatus has retries. At the beginning of the eviction in Updater we call GetStatus. If it returns error or outdated lease, no eviction is made.
- We remove retries from GetStatus. We have single `GetStatus` call in Updater, so it happens that some evictions iterations are skipped due to random cluster errors.
- We remove retries form  GetStatus. We keep checking GetStatus every 10s and sore the result (when there is no error) in Updater. At the beginning of eviction loop we checked component availability based on the last stored status.

I don't see a problem with implementing third solution instead of the first one. I'm not sure if i'm happy with second though...
Wdyt?

Reply:
After offline discussion with @bskiba we decided to keep retries as they are.